### DISCUSSION TITLE
How do airscan and the builtin escl sane backend compare ? Are they compatible ?

### DISCUSSION THREAD

**Discussion by symphorien** [alexpevzner/sane-airscan#322] [General]
Hello,

I'm a downstream SANE maintainer in NixOS, a linux distribution. We had a report from a user that in their specific case they had to disable the builtin escl backend for their scanner to be recognized (by airscan) https://discourse.nixos.org/t/help-with-a-wireless-scanner/36439/17 . I had not really realized that sane has had a builtin escl backend for a few years, and that airscan overlaps with this functionnality.

I wonder:
- how do airscan and the builtin escl backend compare ?
- are they compatible ?
- should we always disable one in favor of the other as a linux distro ? (either in terms of default config, or advice on the wiki)

**Comment by alexpevzner** [comment]
Hi @symphorien,

sane_escl and sane_airscan were developed in the nearly the same time without any knowledge about each other. The very first release of sane_escl was created literally a couple of weeks before sane_airscan and was immediately included into the SANE project due to a high demand on the eSCL driver.

When sane_escl was published, I've looked to it source and decided to continue with sane_airscan. When sane_airscan was released first, I've came to SANE project and proposed to replace sane_escl with sane_airscan. It caused a half-year of hot discussion, and finally I've came to conclusion, that maintaining this project outside of the SANE bundle is more convenient for me and for everybody, so now these projects exist separately.

Meanwhile, sane_airscan has found its way to the most of the Linux distros, as well as to *BSD and ChromeOS, and many distros include sane_airscan and disable sane_escl in the main SANE bundle.

sane_airscan implements everything that sane_escl does excluding the following two things:
* it doesn't support PDF as a transport format, because PDF brings too much external dependencies and I believe doesn't add a lot in a terms of image quality
* it doesn't give access to hardware brightness/contrast control, implementing similar functionality purely in software (which sometimes is not good, depending on hardware)

The following features are unique to sane_airscan and missed in sane_escl:
* WSD support. About 1/3 of supported devices doesn't implement eSCL, only WSD
* intelligent discovery which merges together multiple instances of the same device (sane-escl will return many URLs, sane_airscan will return a single device and will choose between these URLs automatically)
* It doesn't wait in `sane_start` until the entire image is received, so `sane_cancel` works, if supported by device
* in general, I believe sane_airscan implements more workarounds for buggy hardware (i.e., it's compatibility with hardware is better) and is more reliable
* much less memory consumption. sane_escl unpacks the entire image at once. sane_airscan unpacks line by line in most cases. Note, A4 RGB image at 1200 DPI takes about 1/2 Gb of RAM when fully unpacked
* written by professional developer with more that 25 years of experience

The only distro I know that includes both sane_escl and sane_airscan is Debian, due to their policy that this is the user's freedom (and responsibility) to choose.

So hope I was able to answer your question :)
